### PR TITLE
Add simple Pub Sub framework and fix bug during long march

### PR DESCRIPTION
**Pull request by auxten**
- Fix PersistentCaller Reconnect logic
- Make metic collect interval to 1h, the log should be less
- Move client.conn.pCaller init in newConn

### CODE REVIEW

File: chainbus/bus.go [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+
+import (
+	"fmt"
+	"reflect"
+	"sync"
+
+	bi "github.com/CovenantSQL/CovenantSQL/blockproducer/interfaces"
+)
+
+type ChainSuber interface {
```
Comment (covenantdb) [inline chainbus/bus.go:28]: **[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported type ChainSuber should have comment or be unexported

File: chainbus/bus_test.go [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+	if bus.SubscribeOnce(bi.TransactionType(1), "String") == nil {
+		t.Fail()
+	}
+}
+
+func TestSubscribeOnceAndManySubscribe(t *testing.T) {
+	bus := New()
+	event := bi.TransactionType(1)
+	flag := 0
+	fn := func() { flag += 1 }
```
Comment (covenantdb) [inline chainbus/bus_test.go]: **[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
should replace flag += 1 with flag++

File: chainbus/bus_test.go [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+		out <- a
+	}, false)
+
+	bus.Publish(bi.TransactionType(1), 1, results)
+	bus.Publish(bi.TransactionType(1), 2, results)
+
+	var numResults int32
+
+	go func() {
+		for _ = range results {
```
Comment (covenantdb) [inline chainbus/bus_test.go]: **[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
should omit values from range; this loop is equivalent to `for range ...`

**Review by laodouya — APPROVED** [review]
LGTM

**Pull request merged**